server/channel_player.go

```diff
@@ -648,3 +650,93 @@ func (server ChannelServer) playerMoveInventoryItem(conn mnet.Client, reader mpa
 		log.Println(err)
 	}
 }
+
+func (server ChannelServer) playerTakeDamage(conn mnet.Client, reader mpacket.Reader) {
+	mobAttack := reader.ReadInt8()
+	damage := reader.ReadInt32() // Damage amount
+	healSkillID := int32(0)
+
+	if damage < -1 {
+		return
+	}
+
+	reducedDamage := damage
+
+	plr, err := server.players.getFromConn(conn)
+	if err != nil {
+		return
+	}
+
+	if plr.HP() == 0 {
+		return
+	}
+
+	field, ok := server.fields[plr.MapID()]
+	if !ok {
+		return
+	}
+
+	inst, err := field.GetInstance(plr.InstanceID())
+	if err != nil {
+		return
+	}
+
+	var mob mob.Data
+	var mobSkillID, mobSkillLevel byte = 0, 0
+
+	if mobAttack < -1 {
+		mobSkillLevel = reader.ReadByte()
+		mobSkillID = reader.ReadByte()
+	} else {
+		magicElement := int32(0)
+
+		if reader.ReadBool() {
+			magicElement = reader.ReadInt32()
+			_ = magicElement
+			// 0 = no element (Grendel the Really Old, 9001001)
+			// 1 = Ice (Celion? blue, 5120003)
+			// 2 = Lightning (Regular big Sentinel, 3000000)
+			// 3 = Fire (Fire sentinel, 5200002)
+		}
+
+		spawnID := reader.ReadInt32()
+		mobID := reader.ReadInt32()
+
+		mob, err = inst.LifePool().GetMobFromID(spawnID)
+		if err != nil {
+			return
+		}
+
+		if mob.ID() != mobID {
+			return
+		}
+
+		stance := reader.ReadByte()
+
+		reflected := reader.ReadByte()
+
+		reflectAction := byte(0)
+		var reflectX, reflectY int16 = 0, 0
+
+		if reflected > 0 {
+			reflectAction = reader.ReadByte()
+			reflectX, reflectY = reader.ReadInt16(), reader.ReadInt16()
+		}
+
+		playerDamage := -damage
+
+		// Magic guard dmg absorption
+
+		// Fighter / Page power guard
+
+		// Meso guard
+
+		plr.GiveHP(int16(playerDamage))
+
+		inst.Send(player.PlayerReceivedDmg(plr.ID(), mobAttack, damage, reducedDamage, spawnID, mobID, healSkillID, stance, reflectAction, reflected, reflectX, reflectY))
```

Comment:
Could this be wrapped within the player package? Maybe as `GiveHP(dmg int16, show bool)` (might have situations where negative hp might not need to be shown?) as the player has a ptr to the instance they are in. This also keeps the `PlayerReceiveDmg function` in package scope only and gives it the same nomenclature as the other packet functions.

Reply:
Definitely. I will get that setup and amend the PR. I agree that having it scoped to the player package would be more effective and helpful down the line.